src/matchmaking/book/opening_book.cpp

```diff
@@ -42,21 +43,22 @@ void OpeningBook::setup(const std::string& file, FormatType type) {
             throw std::runtime_error("No openings found in EPD file: " + file);
         }
     }
+    if (order_ == OrderType::RANDOM && type != FormatType::NONE){shuffle();}
 }
 
 Opening OpeningBook::fetch() noexcept {
     static uint64_t opening_index = 0;
     const auto idx                = start_ + opening_index++;
     const auto book_size          = std::holds_alternative<epd_book>(book_)
-                                        ? std::get<epd_book>(book_).size()
+                                        ? (std::get<epd_book>(book_).size())
                                         : std::get<pgn_book>(book_).size();
 
     if (book_size == 0) {
         return {chess::constants::STARTPOS, {}};
     }
 
     if (std::holds_alternative<epd_book>(book_)) {
-        const auto fen = std::get<epd_book>(book_)[idx % std::get<epd_book>(book_).size()];
+        const auto fen = std::get<epd_book>(book_)[idx % (std::get<epd_book>(book_).size() - 1)];
```

Comment:
i don't get this..
a) why only subtract for epd books, and not for pgn books as well?
b) idx is "arbitrarily" large, modulo the book size will give an index which is [0, book_size), so why we subtracting one?

Reply:
@Disservin because its too big by 1, i tried debugging the book_.size by printing the function and everytime its larger than the actual opening size by 1, once i reduced it by one the tests no longer stopped prematurely. for the pgn the function outputs the correct amount of openings so i didnt reduce it

Reply:
issue https://github.com/Disservin/fast-chess/issues/262

Reply:
but then the issue is at some other place.. not here this is only a hack

Reply:
yeah, but i dont really know how the book_.size function works to calculate it so that its off by 1, and i checked the output of the pgns to see which openings are used and its correct for this fix, so maybe we can merge this first and do a refactor at some other point? cuz this bug is pretty critical

Reply:
```
        while (safeGetline(openingFile, line)) {
            epd.emplace_back(line);
        }
```
could it be that this part of the code somehow add in an extra line? @Disservin

Reply:
perhaps yea, maybe a 
```cpp
        while (safeGetline(openingFile, line)) {
            if (!line.empty()) epd.emplace_back(line);
        }
```

Reply:
alright ill try it and see the output first

Reply:
fixed it locally for me i think

Reply:
yep fixed it for me too, thank you